python/lsst/cp/pipe/cpSpectroFlat.py

```diff
@@ -0,0 +1,324 @@
+        doc="Filter mode/downstream.",
+    )
+    smoothingOutlierPercentile = pexConfig.Field(
+        dtype=float,
+        default=1.0,
+        doc="Percentile of extreme values to clip from top and bottom of distribution.",
+    )
+
+    # makePixelFlat
+    # photutils.background Background2D, MedianBackground
+    # astropy.stats.SigmaClip
+    doBackgroundRemoval = pexConfig.Field(
+        dtype=bool,
+        default=True,
+        doc="Should amplifier backgrounds be measured and divided out?",
+    )
+    backgroundSigmaClipThreshold = pexConfig.Field(
+        dtype=float,
+        default=3.0,
+        doc="Backgroud sigma clip threshold.",
+    )
+    backgroundBoxSize = pexConfig.Field(
+        dtype=int,
+        default=20,
+        doc="Size of background super-pixel.",
+    )
+    backgroundFilterSize = pexConfig.Field(
+        dtype=int,
+        default=3,
+        doc="Size of smoothing filter for backgrounds.",
+    )
+
+
+class CpSpectroFlatTask(pipeBase.PipelineTask):
+    """Construct a spectro-flat from a white-band flat."""
+
+    ConfigClass = CpSpectroFlatTaskConfig
+    _DefaultName = "cpSpectroFlat"
+
+    def runQuantum(self, butlerQC, inputRefs, outputRefs):
+        inputs = butlerQC.get(inputRefs)
+
+        # Get provenance.  This is essential for debugging.
+        inputMD = inputs['inputFlat'].getMetadata()
+        for inputName in sorted(list(inputs.keys())):
+            reference = getattr(inputRefs, inputName, None)
+            if reference is not None and hasattr(reference, "run"):
+                runKey = f"LSST CALIB RUN {inputName.upper()}"
+                runValue = reference.run
+                idKey = f"LSST CALIB UUID {inputName.upper()}"
+                idValue = str(reference.id)
+
+                inputMD[runKey] = runValue
+                inputMD[idKey] = idValue
+        inputs['inputMD'] = inputMD
+
+        outputs = self.run(**inputs)
+        butlerQC.put(outputs, outputRefs)
+
+    def run(self, dummyExpRef, inputFlat, inputPtc, inputMD):
+        """Create spectro flat from an input flat and PTC.
+
+        Parameters
+        ----------
+        dummyExpRef : `lsst.daf.butler.DeferredDatasetHandle`
+            Exposure used to ensure the correct flat is chosen.
+        inputFlat : `lsst.afw.image.Exposure`
+            Input flat to convert to spectro-flat.
+        inputPtc : `lsst.ip.isr.PhotoTransferCurveDataset`
+            Input PTC containing the gains to apply to spectro-flat.
+        inputMD : `lsst.daf.base.PropertyList`
+            Provenance metadata derived from the butler references.
+
+        Returns
+        -------
+        results : `lsst.pipe.base.Struct`
+            The results struct containing:
+
+            ``outputData``
+                Final modified flat generated from the inputs
+                (`lsst.afw.image.Exposure`).
+        """
+        # Verify input
+        exposureFilter = dummyExpRef.get(component="filter")
```

Comment:
is `dummyExpRef` really needed? It seems that `inputFlat` should be enough.

Reply:
The exposure is needed to add a time constraint on the butler lookup for the flat and PTC.  Without it, there's nothing constraining which calibrations in the input calibration collections to select.